File: samples/run_demo.py
#!/usr/bin/env python3
"""Run demo pipelines from configs/demo/graph/*.json"""

from __future__ import annotations

import sys
import threading
import time
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT))

from simple_pipe.context.frame_meta import FrameInput
from simple_pipe.pipeline.simple_pipeline import SimplePipeline
from simple_pipe.spec.runtime_config import PipelineRuntimeConfig


def run_graph(name: str) -> None:
    graph_path = ROOT / "configs" / "demo" / "graph" / f"{name}.json"
    defaults_dir = ROOT / "configs" / "demo" / "node_defaults"
    pipeline = SimplePipeline(defaults_dir=defaults_dir)

    results: list = []
    done = threading.Event()

    def on_output(meta):
        results.append(meta)
        if len(results) >= 3:
            done.set()

    build_result = pipeline.build_from_json(graph_path)
    if not build_result.ok:
        raise RuntimeError(build_result.error)

    runtime_path = ROOT / "configs" / "demo" / "runtime.example.json"
    if runtime_path.exists():
        pipeline.configure(PipelineRuntimeConfig.load(runtime_path))

    pipeline.set_output_callback(on_output)
    pipeline.start()

    for i in range(3):
        pipeline.push_frame(FrameInput(image=f"frame-{i}", frame_index=i))

    if not done.wait(timeout=5.0):
        raise TimeoutError(f"No output for graph {name}")

    pipeline.stop()
    print(f"[{name}] frames={len(results)} last_safety={results[-1].extensions.get('safety')}")


if __name__ == "__main__":
    run_graph("linear")
    run_graph("safety_fanin")

File: samples/run_demo.cpp
#include <chrono>
#include <iostream>
#include <thread>

#include <nlohmann/json.hpp>

#include "simple_pipe/context/frame_meta.hpp"
#include "simple_pipe/pipeline/simple_pipeline.hpp"

#ifndef SIMPLE_PIPE_SOURCE_DIR
#define SIMPLE_PIPE_SOURCE_DIR "."
#endif

int main() {
  const std::string root = SIMPLE_PIPE_SOURCE_DIR;
  simple_pipe::SimplePipeline linear(root + "/configs/demo/node_defaults");
  if (!(linear.BuildFromJson(root + "/configs/demo/graph/linear.json")).IsSuccess()) {
    std::cerr << "linear build failed\n";
    return 1;
  }

  simple_pipe::SimplePipeline safety;
  if (!(safety.BuildFromJson(root + "/configs/demo/graph/safety_fanin.json")).IsSuccess()) {
    std::cerr << "safety build failed\n";
    return 1;
  }

  int count = 0;
  linear.SetOutputCallback([&](const simple_pipe::FrameMeta& meta) {
    ++count;
    if (count <= 2) {
      std::cout << "[linear] frame=" << meta.frame_index
                << " resize=" << meta.extensions.value("resize", nlohmann::json::object())
                << '\n';
    }
  });
  linear.StartPipeline();
  for (int i = 0; i < 2; ++i) {
    simple_pipe::FrameInput input;
    input.image = "demo";
    input.frame_index = i;
    linear.PushFrame(input);
  }
  std::this_thread::sleep_for(std::chrono::milliseconds(100));
  linear.StopPipeline();

  std::cout << "Demo OK — graphs: linear, safety_fanin (built)\n";
  return 0;
}
